### PR TITLE
[FLINK-25793][connector/base] Implement additive increase/multiplicative decrease to limit throughput

### PR DESCRIPTION
**Pull request by vahmed-hamdy**
## What is the purpose of the change

Introuducing RateLimiting Strategy for `AsyncSinkWriter` base and Additive Increase Multiplicative Decrease Strategy to `KinesisSinkWriter` and `KinesisFirehoseWriter`.

## Brief change log

  - Added `RateLimitingStrategy` Interface and made `AsyncSinkWriter` depend on it.
  - Added `AIMDRateLimitingStrategy` to base module.
  - changed throttling condition in `AsyncSinkWriter` to depend on `RateLimitingStrategy`.
  - modified `KinesisSinkWriter` and `KinesisFirehoseWriter` to depend on `AIMDRateLimitingStrategy`.

## Verifying this change
 - Added unit tests for newly implemented strategy in `flink-connector-base`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AIMDRateLimitingStrategyTest.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    private static final int INCREASE_RATE = 10;
+    private static final int THRESHOLD = 100;
+    private static final double DECREASE_RATE = 0.5;
+
+    @Test
+    public void testInitialRateIsSetByConstructor() {
+        AIMDRateLimitingStrategy rateLimitingStrategy =
+                new AIMDRateLimitingStrategy(INCREASE_RATE, DECREASE_RATE, THRESHOLD, INITIAL_RATE);
+
+        Assertions.assertThat(rateLimitingStrategy.getInFlightMessagesLimit())
```
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AIMDRateLimitingStrategyTest.java:36]: nit: These are more readable with a static import for `assertThat`

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterTest.java [hunk 2]
```diff
@@ -719,6 +721,79 @@ public void testThatFlushingAnEmptyBufferDoesNotResultInErrorOrFailure() throws
         tpts.setCurrentTime(200L);
     }
 
+    @Test
+    public void testThatWriterThrottlesOnRateLimitingStrategy() throws Exception {
+        ExecutorService es = Executors.newFixedThreadPool(10);
```
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterTest.java:726]: Why create a thread pool of size 10 when you are only making 2 requests?

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterTest.java [hunk 3]
```diff
@@ -848,13 +923,34 @@ private void writeTwoElementsAndInterleaveTheNextTwoElements(
...
+            Predicate<AsyncSinkWriterImpl> toAssert)
+            throws IOException, InterruptedException {
+        es.submit(
+                () -> {
+                    batch.forEach(
+                            i -> {
+                                try {
+                                    writer.write(i.toString());
+                                } catch (IOException | InterruptedException e) {
+                                    e.printStackTrace();
```
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterTest.java:939]: If this happens will the test fail?

**Review by dannycranmer — CHANGES_REQUESTED** [review]

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java [hunk 4]
```diff
@@ -229,6 +233,53 @@ public AsyncSinkWriter(
...
+                maxInFlightRequests,
+                maxBufferedRequests,
+                maxBatchSizeInBytes,
+                maxTimeInBufferMS,
+                maxRecordSizeInBytes,
+                new FixedRateLimitingStrategy(maxInFlightRequests * maxBatchSize),
+                states);
+    }
+
+    public AsyncSinkWriter(
```
[the remark this answers is not in the document]
Comment (vahmed-hamdy) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java:250]: I agree, The multiple implementations of constructor are more confusing than helpful.

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java [hunk 5]
```diff
@@ -366,10 +430,19 @@ private void flush() {
      *
      * @param failedRequestEntries requestEntries that need to be retried
      */
-    private void completeRequest(List<RequestEntryT> failedRequestEntries, long requestStartTime) {
+    private void completeRequest(
+            List<RequestEntryT> failedRequestEntries, int messages, long requestStartTime) {
         lastSendTimestamp = requestStartTime;
         ackTime = System.currentTimeMillis();
         inFlightRequestsCount--;
+        inFlightMessages -= messages;
```
[the remark this answers is not in the document]
Comment (dannycranmer) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java:438]: Yep, on second inspection I agree. Thanks

**Comment by fapaul** [comment]
I do not want to be the devil's advocate but this PR looks like a new feature for the AsyncSink that would require a FLIP and we are passed the feature freeze so we can very likely only merge it to 1.16.

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java [hunk 6]
```diff
@@ -229,6 +233,53 @@ public AsyncSinkWriter(
...
+    public AsyncSinkWriter(
+            ElementConverter<InputT, RequestEntryT> elementConverter,
+            Sink.InitContext context,
+            int maxBatchSize,
+            int maxInFlightRequests,
+            int maxBufferedRequests,
+            long maxBatchSizeInBytes,
+            long maxTimeInBufferMS,
+            long maxRecordSizeInBytes,
+            RateLimitingStrategy rateLimitingStrategy,
```
[the remark this answers is not in the document]
Comment (vahmed-hamdy) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java:282]: I agree with wrapping the parameters with a config object, captured in this JIRA [FLINK-26310].
I believe we don't need a builder. The writer should only be created via the sink in normal cases.

**Comment by dannycranmer** [comment]
> I do not want to be the devil's advocate but this PR looks like a new feature for the AsyncSink that would require a FLIP and we are passed the feature freeze so we can very likely only merge it to 1.16.

@fapaul we were treating this change as a bug, however I agree the new interface definitions are edging into new feature territory. During our testing we have found that when the output streams are underprovisioned, saturating the destination stream results in a high number of records throttled. This results in a high retry count and consumes un-necessary CPU/network. See graph below, `aimd8`/`aimd50` contain this fix and `async8`/`async50` do not. You can see that the `aimd` implementation has slightly better throughput and significantly less throttling.

Screenshot 2022-02-23 at 09 59 45

In response to your feedback, in order to remediate the issues observed in KDF/KDS sinks I propose we remove the rate limiting strategy change from the Async Sink base, and move to the concrete implementation. This would leave the Async Sink Base public interface untouched. We can then follow up with a FLIP to define how the rate limiting interfaces will work and pull into the Async Sink base. In order to facilitate this change we would need to extract 2 methods in AsyncSinkWriter that can be overridden to introduce the limiting.

**Comment by dannycranmer** [comment]
I discussed this with @fapaul and we agreed that we can merge this change as a bug fix if we do not modify any public API. Therefore instead of adding the rate limiter to the concrete KDF/KDS implementation we will add it directly to the `AsyncSinkWriter`. We will follow up in Flink 1.16 with a FLIP to consolidate the feature and refactor at this point.

**Comment by fapaul** [comment]
> I discussed this with @fapaul and we agreed that we can merge this change as a bug fix if we do not modify any public API. Therefore instead of adding the rate limiter to the concrete KDF/KDS implementation we will add it directly to the `AsyncSinkWriter`. We will follow up in Flink 1.16 with a FLIP to consolidate the feature and refactor at this point.

Thanks for the detailed analysis and everyone working on this PR. The idea with the automatic buffer size calculations sounds definitely great. Unfortunately, we have to postpone the general availability of configurations a bit to avoid breaking APIs.
Let's work together in 1.16 to roll this out in a configurable way.

**Comment by vahmed-hamdy** [comment]
superseded by #18902

**Pull request closed without merging**